tests/feature_resolution_editions.js

```diff
@@ -423,16 +472,245 @@ tape.test("feature resolution editions precedence", function(test) {
+});
+
+tape.test("feature resolution mixed file features same package", function(test) {
+    var root = protobuf.parse(`edition = "2023";
+    option features.repeated_field_encoding = EXPANDED;
+    message Message1 {
+        repeated int32 expanded = 1;
+        int32 explicit = 2;
+    }`).root;
+    protobuf.parse(`edition = "2023";
+    option features.field_presence = IMPLICIT;
+    message Message2 {
+        repeated int32 packed = 1;
+        int32 implicit = 3;
+    }`, root);
+    root.resolveAll();
+
+    var msg1 = root.lookup("Message1");
+    test.notOk(msg1.fields.expanded.packed, "expanded by default");
+    test.ok(msg1.fields.explicit.hasPresence, "explicit by default");
+
+    var msg2 = root.lookup("Message2");
+    test.ok(msg2.fields.packed.packed, "packed by default");
+    test.notOk(msg2.fields.implicit.hasPresence, "implicit by default");
+
+    test.end();
+});
+
+tape.test("feature resolution mixed file features same package", function(test) {
+    var root = protobuf.parse(`edition = "2023";
+    option features.repeated_field_encoding = EXPANDED;
+    message Message1 {
+        repeated int32 expanded = 1;
+        repeated int32 packed = 2 [features.repeated_field_encoding = PACKED];
+    }`).root;
+    protobuf.parse(`edition = "2023";
+    option features.repeated_field_encoding = PACKED;
+    message Message2 {
+        repeated int32 packed = 1;
+        repeated int32 expanded = 2 [features.repeated_field_encoding = EXPANDED];
+    }`, root);
+    root.resolveAll();
+
+    var expanded = root.lookup("Message1");
+    test.notOk(expanded.fields.expanded.packed, "expanded by default");
+    test.ok(expanded.fields.packed.packed, "packed override");
+
+    var packed = root.lookup("Message2");
+    test.ok(packed.fields.packed.packed, "packed by default");
+    test.notOk(packed.fields.expanded.packed, "expanded override");
+
+    test.end();
+});
+
+tape.test("feature resolution mixed syntax same package", function(test) {
+    var root = protobuf.parse(`syntax = "proto2";
+    message Message2 {
+        optional int32 default = 1;
+        required int32 required = 2;
+        repeated int32 unpacked = 3;
+    }`).root;
+    protobuf.parse(`syntax = "proto3";
+    message Message3 {
+        optional int32 explicit = 1;
+        int32 implicit = 2;
+        repeated int32 packed = 3;
+    }`, root);
+    root.resolveAll();
+
+    var proto2 = root.lookup("Message2");
+    test.ok(proto2.fields.default.hasPresence, "proto2 uses explicit presence");
+    test.ok(proto2.fields.required.required, "proto2 has required fields");
+    test.notOk(proto2.fields.unpacked.packed, "proto2 is expanded by default");
+
+    var proto3 = root.lookup("Message3");
+    test.ok(proto3.fields.explicit.hasPresence, "proto3 optional has explicit presence");
+    test.notOk(proto3.fields.implicit.hasPresence, "proto3 is implicit presence by default");
+    test.ok(proto3.fields.packed.packed, "proto3 is packed by default");
+
```

Comment:
`test.ok(proto3.fields.implicit.required, "proto3 is required by default");`

Reply:
That first test *should* fail.  Proto3 never had "required" fields in the proto2 sense.  Proto3 non-optional fields and proto2 "required" fields are two *very* different things.  Proto3 non-optional fields are "implicit" presence, where if they're unspecified, default values get used.

See https://protobuf.dev/programming-guides/field_presence/ and https://protobuf.dev/best-practices/dos-donts/#dont-add-a-required-field for more information.

Note: this change didn't change proto3 behavior at all.  It only clarified the code and added support for edition 2023.  Your test would have failed before this PR as well, because `required` was simply set `rule === "required"` which was only possible in proto2.

Reply:
It breaks what this PR https://github.com/protobufjs/protobuf.js/pull/2011 introduced

Reply:
https://github.com/Tofandel/protobuf-regression

I'll open an issue as well

Reply:
Responded on the issue, the root problem was that by using the CLI and only upgrading one of your two dependencies, you essentially split this PR in half.  It would be nice if npm had a way of specifying this kind of "optional" dependency, where the latest version of protobufjs should bump your protobufjs-cli dependency automatically